src/storage/src/storage_state.rs

```diff
@@ -502,5 +622,38 @@ impl<'w, A: Allocate> Worker<'w, A> {
         for command in commands {
             self.handle_storage_command(command);
         }
+
+        Ok(())
+    }
+}
+
+struct CommandPump {
+    // TODO(aljoscha): Compute uses a shared `VecDeque` for this. I can change
+    // it to that if we like the basic shape of this. I used a channel just out
+    // of laziness, but we don't need the synchronization/costs that come with
+    // it.
+    command_rx: crossbeam_channel::Receiver<Result<StorageCommand, RecvError>>,
+}
+
+impl CommandPump {
+    fn try_recv(&mut self) -> Result<StorageCommand, TryRecvError> {
+        let wrapped = self.command_rx.try_recv();
+        match wrapped {
+            Ok(Ok(msg)) => Ok(msg),
+            Ok(Err(RecvError)) => Err(TryRecvError::Disconnected),
+            Err(try_recv_error) => Err(try_recv_error),
+        }
+    }
+
+    /// Blocks until a command is available.
+    fn recv<A: Allocate>(&mut self, worker: &mut Worker<A>) -> Result<StorageCommand, RecvError> {
+        while self.is_empty() {
+            worker.timely_worker.step_or_park(None);
```

Comment:
The `VecDeque` can only be appended to from timely operators, so the thread must already have been woken up and done some work. And we read everything from the queue before we park again. In compute, we were already not unparking the thread explicitly, but just activating the operator, which unparks the thread (internally).

For storage, the same will happen now: sending sth on the tokio channel will activate the receiving operator (because it's Activator is in the async Context), and that will unpark the worker thread.

The important thing to note is that the thread that is reading from the queue/crossbeam channel is always the same thread that runs the worker that writes to the queue/channel.

At least that's my reasoning for why/how this is correct.

Reply:
> , so the thread must already have been woken up and done some work.

ah yes, makes sense!